include/nanobind/nb_func.h

```diff
@@ -54,22 +60,61 @@ NB_INLINE PyObject *func_create(Func &&func, Return (*)(Args...),
         nargs = sizeof...(Args);
 
     // Determine the number of nb::arg/nb::arg_v annotations
-    constexpr size_t nargs_provided =
+    static constexpr size_t nargs_provided =
         ((std::is_same_v<arg, Extra> + std::is_same_v<arg_v, Extra>) + ... + 0);
-    constexpr bool is_method_det =
+    static constexpr bool is_method_det =
         (std::is_same_v<is_method, Extra> + ... + 0) != 0;
-    constexpr bool is_getter_det =
+    static constexpr bool is_getter_det =
         (std::is_same_v<is_getter, Extra> + ... + 0) != 0;
+    static constexpr bool has_arg_annotations =
+        nargs_provided > 0 && !is_getter_det;
+
+    static constexpr size_t
+        kwonly_pos_1 = index_1_v<std::is_same_v<kw_only, Extra>...>,
+        kwonly_pos_n = index_1_v<std::is_same_v<kw_only, Extra>...>;
+    // Arguments after nb::args are implicitly keyword-only
+    static constexpr bool explicit_kw_only = kwonly_pos_1 != sizeof...(Extra);
+    static constexpr bool implicit_kw_only = args_pos_1 + 1 < kwargs_pos_1;
 
     // A few compile-time consistency checks
     static_assert(args_pos_1 == args_pos_n && kwargs_pos_1 == kwargs_pos_n,
         "Repeated use of nb::kwargs or nb::args in the function signature!");
-    static_assert(nargs_provided == 0 || nargs_provided + is_method_det == nargs || is_getter_det,
+    static_assert(!has_arg_annotations || nargs_provided + is_method_det == nargs,
         "The number of nb::arg annotations must match the argument count!");
     static_assert(kwargs_pos_1 == nargs || kwargs_pos_1 + 1 == nargs,
         "nb::kwargs must be the last element of the function signature!");
-    static_assert(args_pos_1 == nargs || args_pos_1 + 1 == kwargs_pos_1,
-        "nb::args must follow positional arguments and precede nb::kwargs!");
+    static_assert(args_pos_1 == nargs || args_pos_1 < kwargs_pos_1,
+        "nb::args must precede nb::kwargs if both are present!");
+    static_assert(has_arg_annotations || (!implicit_kw_only && !explicit_kw_only),
+        "Keyword-only arguments must have names!");
+
+    // Avoid instantiating count_args_before_index unless the kw_only
+    // annotation is actually used
+    static constexpr size_t nargs_before_kw_only = []() constexpr {
```

Comment:
Not a fan of this one, it seems kind of complicated. I am suspicious of what a compiler like MSVC will make out of this (I've seen it struggle with things as basic as inlining a 1-line lambda function, and it is generally bad at constexpr computation).

Reply:
Switched to a version that doesn't use a lambda.